sdk/core/azure-core-http-httpurlconnection/src/main/java/com/azure/core/http/httpurlconnection/SocketClient.java

```diff
@@ -0,0 +1,148 @@
+
+            HttpUrlConnectionResponse response = buildResponse(httpRequest, in);
+
+            String redirectLocation = response.getHeaders().stream()
+                .filter(h -> h.getName().equals("Location"))
+                .map(HttpHeader::getValue)
+                .findFirst()
+                .orElse(null);
+
+            if (redirectLocation != null) {
+                httpRequest.setUrl(redirectLocation);
+                return sendPatchRequest(httpRequest);
+            }
+            return response;
+        }
+    }
+
+    /**
+     * Converts an instance of HttpRequest to a String representation for sending
+     * over the output stream
+     *
+     * @param httpRequest {@link com.azure.core.http.HttpRequest} instance
+     * @return the String representation of the HttpRequest
+     */
+    private static String buildPatchRequest(HttpRequest httpRequest) {
+        final StringBuilder request = new StringBuilder();
+
+        request.append("PATCH")
+            .append(" ")
+            .append(httpRequest.getUrl().getPath())
+            .append(HTTP_VERSION)
+            .append("\r\n");
+
+        if (httpRequest.getHeaders().getSize() > 0) {
+            for (HttpHeader headerLine : httpRequest.getHeaders()) {
+                request.append(headerLine.getName())
+                    .append(": ")
+                    .append(headerLine.getValue())
+                    .append("\r\n");
+            }
+        }
+        // Add the body if there is a body to add
+        if (httpRequest.getBody() != null) {
+            request.append("\r\n")
+                .append(httpRequest.getBodyAsBinaryData().toString())
+                .append("\r\n");
+        }
+        request.append("\r\n");
+        return request.toString();
+    }
+
+    /**
+     * Reads the response from the input stream and extracts the information
+     * needed to construct an instance of HttpUrlConnectionResponse
+     *
+     * @param httpRequest {@link com.azure.core.http.HttpRequest} instance
+     * @param reader {@link java.io.BufferedReader} instance
+     * @return an instance of HttpUrlConnectionResponse
+     */
+    private static HttpUrlConnectionResponse buildResponse(HttpRequest httpRequest, BufferedReader reader) throws IOException {
+        String statusLine = reader.readLine();
+        int dotIndex = statusLine.indexOf('.');
+        int statusCode = Integer.parseInt(statusLine.substring(dotIndex+3, dotIndex+6));
+
+        Map<String, List<String>> headers = new HashMap<>();
+        String line;
+        while ((line = reader.readLine()) != null && !line.isEmpty()) {
+            String[] kv = line.split(": ", 2);
+            String k = kv[0];
+            String v = kv[1];
+            headers.computeIfAbsent(k, key -> new ArrayList<>()).add(v);
+        }
+
+        StringBuilder bodyString = new StringBuilder();
+        while ((line = reader.readLine()) != null) {
+            bodyString.append(line).append("\n");
```

Comment:
Could this introduce a bug where there is an additional trailing `\n` in the response body? Also, isn't there a newline between the headers and response body?